modules/apps/headless-apio/structured-content/structured-content-apio-test/src/testIntegration/java/com/liferay/structured/content/apio/internal/architect/resource/test/BaseStructuredContentNestedCollectionResourceTestCase.java

```diff
@@ -42,6 +45,64 @@
  */
 public abstract class BaseStructuredContentNestedCollectionResourceTestCase {
 
+	protected String getDDMFormFieldDataType(
+			DDMFormFieldValue ddmFormFieldValue, DDMStructure ddmStructure)
+		throws Exception {
+
+		NestedCollectionResource nestedCollectionResource =
+			_getNestedCollectionResource();
+
+		Class<? extends NestedCollectionResource> clazz =
+			nestedCollectionResource.getClass();
+
+		Class<?>[] declaredClasses = clazz.getDeclaredClasses();
+
+		Class<?> innerClass = declaredClasses[0];
+
+		Constructor constructor = innerClass.getDeclaredConstructor(
+			DDMFormFieldValue.class, DDMStructure.class);
+
+		constructor.setAccessible(true);
+
+		Object object = constructor.newInstance(
+			ddmFormFieldValue, ddmStructure);
+
+		Method method = innerClass.getMethod("getDDMFormFieldDataType");
+
+		method.setAccessible(true);
+
+		return (String)method.invoke(object);
+	}
+
+	protected String getDDMFormFieldInputControl(
+			DDMFormFieldValue ddmFormFieldValue, DDMStructure ddmStructure)
+		throws Exception {
+
+		NestedCollectionResource nestedCollectionResource =
+			_getNestedCollectionResource();
+
+		Class<? extends NestedCollectionResource> clazz =
+			nestedCollectionResource.getClass();
+
+		Class<?>[] declaredClasses = clazz.getDeclaredClasses();
+
+		Class<?> innerClass = declaredClasses[0];
+
+		Constructor constructor = innerClass.getDeclaredConstructor(
+			DDMFormFieldValue.class, DDMStructure.class);
+
+		constructor.setAccessible(true);
+
+		Object object = constructor.newInstance(
+			ddmFormFieldValue, ddmStructure);
+
+		Method method = innerClass.getMethod("getDDMFormFieldInputControl");
```

Comment:
You can extract this, right? Or are you following the rule of 3 in duplication?

Reply:
Indeed, avoiding premature refactoring